pipelines/optimus/Optimus.wdl

```diff
@@ -12,34 +12,60 @@ import "AlignTagCorrectUmis.wdl" as AlignTagCorrectUmis
 # format.
 workflow Optimus {
 
-  Array[Array[File]] fastq_inputs  # array of arrays of matched fastqs [ [r1, r2, i1], ... ]
-  File whitelist  # 10x genomics cell barcode whitelist for 10x V2
+  # Sequencing data inputs (fastq)
+  Array[File] r1 # forward read, contains cell barcodes and molecule barcodes
+  Array[File] r2 # reverse read, contains cDNA fragment generated from captured mRNA
+  Array[File]? i1 # (optional) index read, for demultiplexing of multiple samples on one flow cell.
+
+  String sample_id  # name of sample matching this file, inserted into read group header
+
+  # organism reference parameters
   File tar_star_reference  # star reference
   File annotations_gtf  # gtf containing annotations for gene tagging
   File ref_genome_fasta  # genome fasta file
-  String sample_id  # name of sample matching this file, inserted into read group header
+
+  # 10x v2 parameters
+  File whitelist  # 10x genomics cell barcode whitelist for 10x V2
+
+  # environment-specific parameters
   String fastq_suffix = ""  # when running in green box, need to add ".gz" for picard to detect
+  Array[Int] indices = range(length(r1)) # this scatters matched [r1, r2, i1] fastq arrays
 
-  # this scatters matched [r1, r2, i1] fastq arrays
-  scatter (fastqs in fastq_inputs) {
+  scatter (index in indices) {
     call fq2bam.FastqToUBam {
       input:
-        fastq_file = fastqs[1],
+        fastq_file = r2[index],
         sample_id = sample_id,
         fastq_suffix = fastq_suffix
     }
 
-    call attach.Attach10xBarcodes {
-      input:
-        r1 = fastqs[0],
-        i1 = fastqs[2],
-        u2 = FastqToUBam.bam_output,
-        whitelist = whitelist
+    # if the index is passed, attach it to the bam file
+    if (defined(i1)) {
+      Array[File] non_optional_i1 = select_first([i1])
+      call attach.Attach10xBarcodes as attach_barcodes {
+        input:
+          r1 = r1[index],
+          i1 = non_optional_i1[index],
+          u2 = FastqToUBam.bam_output,
+          whitelist = whitelist
+      }
     }
 
+    # if the index is not passed, proceed without it.
+    if (!defined(i1)) {
```

Comment:
I couldn't think of a better way to do this 😢

Reply:
👍

Reply:
i guess if we 100% sure we won't need I1,we can just simple take I1 out of the pipeline.